spatial/transform/umeyama.go

```diff
@@ -0,0 +1,228 @@
+	covXY := mat.NewDense(u.m, u.m, nil)
+	covXY.Mul(yc.T(), xc)
+	covXY.Scale(1/float64(u.n), covXY)
+
+	// Singular Value Decomposition
+	var svd mat.SVD
+	if !svd.Factorize(covXY, mat.SVDFull) {
+		return 0, nil, nil, mat.ErrFailedSVD
+	}
+
+	// Get U and V.
+	var uu, v mat.Dense
+	svd.UTo(&uu)
+	svd.VTo(&v)
+
+	// Create identity matrix.
+	s := mat.NewDiagDense(u.m, nil)
+	for i := 0; i < u.m; i++ {
+		s.SetDiag(i, 1)
+	}
+
+	// Check determinants to ensure proper rotation matrix (not reflection).
+	if mat.Det(&uu)*mat.Det(&v) < 0 {
+		s.SetDiag(u.m-1, -1)
+	}
+
+	// Calculate scale factor c.
+	singularValues := svd.Values(nil)
+	for i := 0; i < u.m; i++ {
+		c += singularValues[i] * s.At(i, i)
+	}
+	c /= u.varX
+
+	// Calculate rotation matrix R.
+	r = mat.NewDense(u.m, u.m, nil)
+	r.Product(&uu, s, v.T())
+
+	// Calculate translation vector t.
+	t = mat.NewVecDense(u.m, nil)
+	rMuX := mat.NewVecDense(u.m, nil)
+	rMuX.MulVec(r, u.muX)
+
+	t.CopyVec(u.muY)
+	t.AddScaledVec(t, -c, rMuX)
+
+	return c, r, t, nil
+}
+
+// UmeyamaTransform finds the similarity transformation between two sets of points
+// that minimizes the mean squared error between them.
+//
+// The transformation relates two sets of n corresponding points {x_i}
+// and {y_i} as:
+//
+//	y_i ≈ c * R * x_i + t,  i=0,...,n-1
+//
+// where c is the scale factor, R is the rotation matrix and t is
+// the translation vector.
+//
+// The point sets are represented as two n×m matrices X and Y, where
+// m is the number of dimensions and x_i and y_i are stored in the i-th
+// row of X and Y, respectively. Typically, m is equal to 2 or 3.
+// If the dimensions of X and Y are not equal, UmeyamaTransform will panic.
+//
+// UmeyamaTransform returns the scale factor c, the rotation matrix R and the translation
+// vector t.
+//
+// If the required SVD fails, UmeyamaTransform will return a mat.ErrFailedSVD.
+//
+// UmeyamaTransform automatically checks for degenerate input by comparing the variance
+// of x with machine epsilon. This is necessary because a variance equal or close
+// to zero may cause numerical instability and/or division by zero.
+// In case of variance ≤ machine epsilon, UmeyamaTransform will return a DegenerateInputError.
+//
+// For more control over variance checking, use NewUmeyama to create an Umeyama
+// instance, inspect its variance with the Var method, and call Transform if appropriate.
+//
+// Reference:
+// "Least-Squares Estimation of Transformation Parameters Between Two Point Patterns"
+// by Shinji Umeyama, IEEE Transactions on Pattern Analysis and Machine Intelligence,
+// Vol. 13, No. 4, April 1991, [doi:10.1109/34.88573].
+// [doi:10.1109/34.88573]: https://doi.org/10.1109/34.88573
+func UmeyamaTransform(x, y *mat.Dense) (c float64, r *mat.Dense, t *mat.VecDense, err error) {
+	u := NewUmeyama(x, y)
+
+	if u.varX <= math.Nextafter(1.0, 2.0)-1.0 {
+		return 0, nil, nil, DegenerateInputError(u.varX)
+	}
+
+	return u.Transform()
```

Comment:
We do need the error here, but I think it should be local to the package. I don't think we need to say anything about SVD; ` var ErrFactorizationFailed = errors.New("transform: factorization failed")`.